interp/src/primitives.rs

```diff
@@ -27,114 +32,191 @@ pub fn update_cell_state(
 
     match cell_type.id.as_str() {
         "std_reg" => {
+            // TODO: this is wrong...
             let write_en = ir::Id::from("write_en");
 
             // register's write_en must be high to write reg.out and reg.done
-            if new_env.get(&cell, &write_en) != 0 {
+            if new_env.get(&cid, &cell, &write_en) != 0 {
                 let out = ir::Id::from("out"); //assuming reg.in = cell.out, always
                 let inp = ir::Id::from("in"); //assuming reg.in = cell.out, always
                 let done = ir::Id::from("done"); //done id
 
-                new_env.put(cell, &output[0], env.get(&inputs[0], &out)); //reg.in = cell.out; should this be in init?
+                new_env.put(
+                    &cid,
+                    cell,
+                    &output[0],
+                    env.get(&cid, &inputs[0], &out),
+                ); //reg.in = cell.out; should this be in init?
 
                 if output[0].id == "in" {
-                    new_env.put(cell, &out, new_env.get(cell, &inp)); // reg.out = reg.in
-                    new_env.put(cell, &done, 1); // reg.done = 1'd1
-                    new_env.remove_update(cell); // remove from update queue
+                    new_env.put(
+                        &cid,
+                        cell,
+                        &out,
+                        new_env.get(&cid, cell, &inp),
+                    ); // reg.out = reg.in
+                    new_env.put(&cid, cell, &done, 1); // reg.done = 1'd1
+                                                       //new_env.remove_update(cell); // remove from update queue
                 }
             }
         }
+        "std_mem_d1" => {
```

Comment:
A note to think about: do we need to treat memories differently than other cells? Certainly they keep state beyond just the values on their ports.